add rep-0152.rst for orphaned package maintainer info. (#150)
ros-infrastructure/rep: rep-0152.rst

REP: 152
Title: Orphaned Package Maintaining
Author: Kei Okada
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 27-Dec-2017

Outline
=======

#. Abstract_
#. Motivation_
#. Reference_
#. Operations_
#. References_
#. Copyright_

Abstract
========

This REP specifies how we can organize and release packages without active maintainers.

Motivation
==========

> As the size of the ROS ecosystem grows, there are several packages which losing active maintainers.

Reviewer: ```suggestion
As the size of the ROS ecosystem grows, some packages will lose active maintainers.
```

> The main drawback of this situation is these packages do not get released into newer ROS distro coming every year, even if they are commonly used (For example, USB camera drivers are not released into Kinetic for more than half year [1]_ [2]_).

Reviewer: ```suggestion
The main drawback of this situation is these packages do not get released into newer ROS distros which are released every year, even if they are commonly used (For example, USB camera drivers have not been released into Kinetic for more than half year [1]_ [2]_).
```

> The main drawback of this situation is these packages do not get released into newer ROS distros which are released every year, even if they are commonly used (For example, USB camera drivers have not been released into Kinetic for more than half year [1]_ [2]_).

Reviewer 2: ```suggestion
The main drawback of this situation is these packages do not get released into newer ROS distros, which are released every year, even if they are commonly used (For example, USB camera drivers have not been released into Kinetic for more than half year [1]_ [2]_).
```

> Ideally everyone willing to use newer ROS distro almost equivalent to the set of packages in previous ROS distro.

Reviewer: ```suggestion
Ideally, everyone wants to use a newer ROS distro which contains a set of packages almost equivalent to the set of packages in the previous ROS distro.
```

> It is understandable that original maintainer could not continue serving as an active maintainer, because of starting a new project, moving an institute/company, losing interest and so on.

Reviewer: ```suggestion
It is understandable that the original maintainer of a package might not continue serving as an active maintainer due to starting a new project, moving institutions/companies, losing interest, and so on.
```

> And being a maintainer is not an easy job.

Reviewer: ```suggestion
Being a maintainer is not an easy job.
```

> It includes responding to incoming questions and issues, check for proposed changes and decide to add them to the main source tree and so on.

Reviewer: ```suggestion
It includes responding to incoming questions and issues, checking for proposed changes and deciding to add them to the main source tree, and other responsibilities.
```

> The idea behind this document is to have a team of volunteers who can do ONLY the release job, which includes sending source tree to ros build firm and act on notification mail form that.

Reviewer: ```suggestion
The idea behind this document is to organize a team of volunteers who can do ONLY the release job, which includes sending the source tree to the ROS Build Farm and acting on notification emails form the builds.
```

> The volunteers of release maintainers will enable the community to continue using the same set of previous ROS distros with a minimum set of community efforts.

Reviewer: ```suggestion
The volunteers of the release maintainers group will enable the community to continue using the same set of packages from previous ROS distros with minimal community effort.
```

> To this end, we followed Debian QA Group System and Debian Developer's Reference ('Dealing with inactive and/or unreachable maintainers'[3]_), which is considered non-normative, but best practice.

Reviewer: ```suggestion
To this end, we followed the Debian QA Group system and the Debian Developer's Reference ('Dealing with inactive and/or unreachable maintainers'[3]_), which is considered non-normative, but best practice.
```

> In ROS system, we use 'Unmaintained' package status([4]_) and ROS Orphaned Packages Group([5]_), who do not have to take full responsibility to package maintenance jobs but expecting to take minimum responsibilities such as just to releasing into newer ROS distros.

Reviewer: ```suggestion
In ROS, we use 'Unmaintained' package status([4]_) and ROS Orphaned Packages Group([5]_), who do not have to take full responsibility for package maintenance jobs but are expected to take minimum responsibilities which mostly include releasing their packages into newer ROS distros.
```

Reference
=========

REP-0140 [6]_ and REP-0141 [7]_ provide maintainer information in package manifest file and maintainer status in ROS distribution file.

Maintenance Guide in ROSWiki [8]_ provides how to claim maintainership for orphaned packages.

Operations
==========

1-A. Orphaning a package by maintainers
---------------------------------------

If you can no longer maintain your package.

a. Set the package maintainer section in the ``package.xml`` file as follows.

   ``<maintainer email="ros-orphaned-packages@googlegroups.com">ROS Orphaned Package Maintainers</maintainer>``

b. Post a message to the `ROS Discourse Release Category <http://discourse.ros.org/c/release>`_ indicating that package is now orphaned.
   Wait for response from someone in the ROS Orphaned Package Maintainer team.

> c. Add admin permission of both source and release repository to who respond from in the ROS Orphaned Package Maintainers team.

Reviewer 3: ```suggestion
c. Add admin permission of both source and release repository to whomever responded from the ROS Orphaned Package Maintainers team.
```

Reviewer 3: Also, what happens if the maintainer never gets a response?  That's not ideal, obviously, but in the case it does happen, what should the maintainer do?

Relatedly, how long should they wait for a response?  A week?  A month?

Author: If that happens, it means Ophaned team/framework is not working. I think that's is out of the scope of this document. how do you think?

1-B. Orphaning a package by users
---------------------------------

> If you find a package is not actively maintained and willing to release as a ROS Orphaned Package Maintainer.

Reviewer: ```suggestion
If you find a package is not actively maintained and are willing to release as a ROS Orphaned Package Maintainer.
```

a. Contact the current maintainer, ask for a transition to orphaned package and request access to the source and release repository to make releases.

> b. If the current maintainer is not responsible (a single day isn't enough to conclude that) and you still want to release the packages, there are two different paths forward depend on the "location" of the source and release repositories.

Reviewer: ```suggestion
b. If the current maintainer is not responsive (a single day isn't enough to conclude that) and you still want to release the packages, there are two different paths forward depending on the "location" of the source and release repositories.
```

> b. If the current maintainer is not responsive (a single day isn't enough to conclude that) and you still want to release the packages, there are two different paths forward depending on the "location" of the source and release repositories.

Reviewer 3: This says that a single day isn't enough, but some guidance on how long would be appropriate.  I'll suggest 3 months is sufficient time, though it is pretty long if a user is waiting for a package.

Author: how about 'We are asking users to be patient and wait for 3 months at most.'

   1. If the repositories are hosted on GitHub within a `ros-` org unit, the admins of the org group will usually grant the necessary access to these repositories.

>    2. If nobody else besides the original author (which doesn't respond) has access, you should fork the repository.

Reviewer: ```suggestion
   2. If no one else aside from the original author (who doesn't respond) has access, you should fork the repository.
```

>    1. If the repositories are hosted on GitHub within a `ros-` org unit, the admins of the org group will usually grant the necessary access to these repositories.

Reviewer 3: This is what gets done in the end, but it doesn't tell the interested party what to do.  In other words, I think we need a decision tree here which is something like:

* Contact the maintainer.
* Maintainer responded? ----> done
* Maintainer didn't respond after 3 months?
    * Send information to ros-orphaned-packages@googlegroups.com about the source repository, release repository, and where and how maintainer failed to respond.
    * If the ROS Orphaned Package Maintainers agree with your assessment, and the repository is hosted on GitHub within a `ros-` organization, the admins of the group can grant the necessary access.
    * If no one aside from the original author has access, you should fork the repository

Author: thanks for the comment, I have uploaded a fixed version. As for the last comment, do you mean we'd add something like
![diagram](https://user-images.githubusercontent.com/493276/108445992-b875e500-72a0-11eb-875f-82d51677dd42.png) ???

Reviewer 3: Yeah, that is kind of the thing I was thinking of.  Basically a clear set of steps that a person needs to go through in order to take over a package.

Author: @clalancette please check the diagram at https://github.com/ros-infrastructure/rep/pull/150#discussion_r587199922, which includes all processes.

c. Add ROS Orphaned Package Maintainers to the maintainer section of package.xml as follows.

   ``<maintainer email="ros-orphaned-packages@googlegroups.com">ROS Orphaned Package Maintainers</maintainer>``

2. Adopting packages
--------------------

Orphaned packages are seeking a new maintainer.

> If you are willing to step forward the package maintainer, more than to release into new ROS distros, but respond to the issues, pull requests and so on.

Reviewer: ```suggestion
If you are willing to step forward as an official package maintainer (not just releasing into new ROS distros but also respond to issues, pull requests and so on):
```

> Contact to the ROS Orphaned Package Maintainers by posting a message to the `ROS Discourse Release Category <http://discourse.ros.org/c/release>`_.

Reviewer 3: ```suggestion
Contact the ROS Orphaned Package Maintainers by posting a message to the `ROS Discourse Release Category <http://discourse.ros.org/c/release>`_.
```

> Contact the ROS Orphaned Package Maintainers by posting a message to the `ROS Discourse Release Category <http://discourse.ros.org/c/release>`_.

Reviewer 4: Are the orphan maintainers the right body to determine succession eligibility? That's a very different role than volunteer releasing. Currently this is not clearly defined and is the responsibility of the rosdistro release manager.

What is the process for adoption beyond contact? We should have a full cycle defined to not end up with the orphaned packages list growing unbounded.

Author: @tfoote The role of 'maintainer' is described at https://github.com/ros-infrastructure/rep/pull/150/files#diff-ad591b0fb51969339c5dbf52c2336309210bb7b1c89563afff9daca93bd2b911R31-R33 . I understand your concern, one idea is to hold an orphaned maintainer meeting and make a vote to judge that.
however in reality, due to the limited resource of an orphaned maintainer team, I think we will pass maintainership to everyone who requested. Even though we do not sure if they have enough resources/potential to maintain that. If that happens, the package simply goes back to the orphaned team.

Reviewer 4: Yes, that's exactly my concern. You're patterning this on the debian model. However they have a very strong vetting process before someone is allowed to become a debian maintainer, with a process including sponsors and mentorship. We have been relying on direct succession planning and maintainer handoffs and adhoc reviews by the release managers.  We cannot simply handoff maintainership to the first person who requests to do so even if they are not qualified. 

This is exactly why I'm asking about the full cycle. We need to make sure that the whole process is sustainable and not just kicking the can down the road and putting an unsustainable burden on one or a group of people which will box us further into a tighter corner.

Author: @tfoote I totally agree on this. This document says "If you want to maintain a package, please post to ROS Discourse". But did not say how to deal with that request.

I am expecting some discussion will occur on Discourse and decide whether if we could pass the maintainership. 
So how about adding 
```
To pass maintainership to volunteers, you need at least two approvals from the ROS Orphaned Package Maintainers team. And a single day isn't enough to decide new maintainers, the team will wait for other candidates and make the decision for 3 months for at least".
```
??

I understand this proposal is a simplified debian model, and if we could have the strict process as debian, that would be great. I also share your concern about passing maintainership to someone who is not responsible for that.  But to build a sustainable model that fits the current ROS userbase, I think passing maintainership as much as possible and increase the number of potential maintainers, I wouldn't say that we can pass maintainership to the first responders. Even if they failed to keep maintaining, we, the orphaned maintainer team, will backup the process.

So my suggestion here is to not go through the cycle step by step very carefully like debian, we choose to turn the cycle a little faster, (not as much as fast), to build up this orphaned package maintainer community. 
![diagram](https://user-images.githubusercontent.com/493276/109922023-19210b00-7d00-11eb-91c7-02495b1dc711.png)

3. Releasing orphaned packages
------------------------------

> a. The package release process is as same as standard packages, so please follow bloom tutorial titled ``Releasing into a new ROS Distribution`` in ROS wiki documents [9]_ to release the orphaned package.

Reviewer: ```suggestion
a. The package release process is the same as for standard packages, so please follow bloom tutorial titled ``Releasing into a new ROS Distribution`` in the ROS wiki documents [9]_ to release the orphaned package.
```

>    1. Updated maintainer section in ``package.xml`` indicating that the package is orphaned and commit to the source tree before you run the ``catkin_generate_changelog`` command.

Reviewer: ```suggestion
   1. Update the maintainer section in the ``package.xml`` file indicating that the package is orphaned and commit to the source tree before you run the ``catkin_generate_changelog`` command.
```

>    2. During ``bloom-release`` process, you will be asked to set maintenance status as follows.

Reviewer: ```suggestion
   2. During ``bloom-release`` process, you will be asked to set maintenance status.
```

   Choose ``unmaintained`` . ::
      
::

  Would you like to add a maintenance status for this repository?' [Y/n] Y
   Please enter a maintenance status.
   Valid maintenance statuses:
   - developed: active development is in progress
   - maintained: no new development, but bug fixes and pull requests are addressed
   - unmaintained: looking for new maintainer, bug fixes and pull requests will not be addressed
   - end-of-life: should not be used, will disappear at some point

b. Add your package to the list of orphaned packages in `ROS OrphanedPackage page <http://wiki.ros.org/OrphanedPackage>`_ of ROSWiki.

4. Joining ROS Orphaned Package Maintainers
-------------------------------------------

> To join ROS Orphaned Package Maintainers, please subscribe ros-orphaned-packages@googlegroups.com mailing list and also add your name and GitHub account to `ROS OrphanedPackage page <http://wiki.ros.org/OrphanedPackage>`_ .

Reviewer: ```suggestion
To join ROS Orphaned Package Maintainers, please subscribe to the ros-orphaned-packages@googlegroups.com mailing list and also add your name and GitHub account to the `ROS OrphanedPackage page <http://wiki.ros.org/OrphanedPackage>`_ .
```

Title: Orphaned Pakcage Maintaining
Author: Kei Okada
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 27-Dec-2017

Outline
=======

#. Abstract_
#. Motivation_
#. Rationale_
#. Operations_
#. References_
#. Copyright_

Abstract
========

This REP specifies how we can organize and release packages without active maintainers.

Motivation
==========

As the size of ROS ecosystem grows, there are several packages which loosing active maintainers.
Main drawback of this situation is these packages do not get released into newer ROS distro coming every year, even if they are commonly used (For example, usb camera drivers are not released into Kinetic for more than half year [1]_ [2]_).
Ideally everyone willing to use newer ROS distro almost equivalent to the set of packages in previous ROS distro.

It is understandable that original maintainer could not continue serving as an active maintainer, because of starting new project, moving an institute/company, loosing interest and so on.
And being a maintainer is not a easy job. It includes respond to incoming questions and issues, check for proposed changes and decide to include them to a main source tree and so on.

> The idea behind this document is to have a team of volunteers who can do ONLY the release job.

Reviewer 5: I think it is crucial to mention that when doing a release of a package that person *must* also act on potential notification emails from the buildfarm. That implies to set a maintainer email which the person receives (as mentioned already). But I think the expectation to act on notification emails from the buildfarm should be explicitly stated - otherwise that burden falls onto the people operating the buildfarm which isn't acceptable imo.

Reviewer 5: Sounds good to me.

This will enable to continue using same set of previous ROS distros with minimum set of community efforts.
To this end, we followed Debian QA Group System and Debian Developer's Reference ('Dealing with inactive and/or unreachable maintainers'[3]_), which is considered non-normative, but best practice.

In ROS system, we use 'Unmaintained' package status([4]_) and ROS Orphaned Packages Group([5]_), who do not have to take full responsibility to package maintenance jobs, but expecting to take minimum responsibilities such as just to releasing into newer ROS distros.

Rationale
=========

REP-0140 [6]_ and REP-0141 [7]_ provide maintainer information in package manifest file and maintainer status in ROS distribution file.

Maintenance Guide in ROSWiki [8]_ provides how to claim maintainership for orphaned packages.

Operations
==========

> 1-A. Orphaning a package by maintainers

Reviewer 5: With the current proposal I don't see a way to distinguish the following two cases:

* The group of orphaned package maintainers "chooses" that a specific package is orphaned, updates the manifest accordingly and starts doing releases of that package
* A maintainer who can no longer maintain a package updates the manifest accordingly.

From a users point of view it would be nice to be able to distinguish these two cases. In the first case there is the expectation that there will be releases in the future. In the second case likely not (until the orphaned package maintainers actually "accept" the work and start doing releases for that package).

Author: @dirk-thomas thank you for comments,  i have fixed most of comment. On this comment, `A. Orphaned a package by maintainer` corresponds to 
`A maintainer who can no longer maintain a package updates the manifest accordingly`.
 and if some of maintainer want to maintain the package by the orphaned team, I expect they choose `Orphaning a package by users`, not as a member of the team, but as an individuals.

To minimize the effort of orphaned maintainers, I intentionally remove the process of something like `the group of orphaned package maintainers "chooses" `So in both cases, user might be not sure if the packages will be releases in the future.  To put such responsibility to the orphaned team feel to much for me. How do you think?

Reviewer 5: So for all packages marked as "orphaned" in the manifest files there is no information which ones are going to be released by the group, right? I think it would be valuable for users to differentiate them but I understand your rational why the group doesn't want to "promise" that.

Author: Yeah, my first thought on this, is to create the organization like `ros-unmaintained` and once the package is marked as orphaned, then we transfer the package to that organization. If so everyone can see who is going to (or able to ) release these package, by looking at the "users" or "teams" of the organization. 
But the draw back of this idea is that the location of the package will be moving around... If we can create github teams across different organization, then it would be great.
c.f. https://discourse.ros.org/t/releasing-repositories-form-other-people/1797/11?u=k-okada

Reviewer 5: > the draw back of this idea is that the location of the package will be moving around

:-1: on moving repos around just to indicate their status. It generates quite some friction which is unnecessary.

b. If the current maintainer is not responsible (a single day isn't enough to conclude that) and you still want to release the packages, there are two different paths forward depend on the "location" of the source and release repositories.

   1. If the repositories are hosted on GitHub within an `ros-` org unit, the admins of the org unit will usually grant the necessary access to these repositories.

   2. If nobody else beside the original author (which doesn't respond) has access, you should fork the repository.

c. Add ROS Orphaned Package Maintainers to the maintainer section of package.xml as follows.

``<maintainer email="ros-orphaned-packages@googlegroups.com">ROS Orphaned Package Maintainers</maintainer>``

2. Releasing orphaned packages
------------------------------

a. The package release process is as same as normal packages, so please follow bloom tutorial titled ``Releasing a Package for the First Time`` in ROS wiki documents [9]_ to release the orphaned package.

   1. Updated maintainer section in ``package.xml`` indicating that the package is orphaned and commit to the source tree, before you run the ``catkin_generate_changelog`` command.

>    2. During ``bloom-release`` process, you will asked to set maintenance status as follows. Choose ``unmaintained`` . ::

Reviewer 5: Nitpick: wrap before new sentence.

      
::

  Would you like to add a maintenance status for this repository?' [Y/n] Y
   Please enter a maintenance status.
   Valid maintenance statuses:
   - developed: active development is in progress
   - maintained: no new development, but bug fixes and pull requests are addressed
   - unmaintained: looking for new maintainer, bug fixes and pull requests will not be addressed
   - end-of-life: should not be used, will disappear at some point

b. Add your package to the list of orphaned packages in `ROS OrphanedPackage page <http://wiki.ros.org/OrphanedPackage>`_ of ROSwiki.

> 3. Adopting packages

Reviewer 5: I think this should be mentioned before `2. Releasing orphaned packages` since it is clearly the preferred solution.

--------------------

> Orphaned packages are seeking a new maintainer. If you willing to step forward the package maintainer, more than just to release into new ROS distros, but respond to the issues, pull requests and so on.

Reviewer 5: Nitpick: wrap before new sentence.

It is understandable that original maintainer could not continue serving as an active maintainer, because of starting new project, moving an institute/company, loosing interest and so on.
And being a maintainer is not a easy job.
It includes respond to incoming questions and issues, check for proposed changes and decide to include them to a main source tree and so on.

The idea behind this document is to have a team of volunteers who can do ONLY the release job, which includes send source tree to ros build firm and act on notification mail form that.
This will enable to continue using same set of previous ROS distros with minimum set of community efforts.
To this end, we followed Debian QA Group System and Debian Developer's Reference ('Dealing with inactive and/or unreachable maintainers'[3]_), which is considered non-normative, but best practice.

In ROS system, we use 'Unmaintained' package status([4]_) and ROS Orphaned Packages Group([5]_), who do not have to take full responsibility to package maintenance jobs, but expecting to take minimum responsibilities such as just to releasing into newer ROS distros.

Rationale
=========

> REP-0140 [6]_ and REP-0141 [7]_ provide maintainer information in package manifest file and maintainer status in ROS distribution file.

Reviewer 6: This is a reference, not rationale.

> Maintenance Guide in ROSWiki [8]_ provides how to claim maintainership for orphaned packages.

Reviewer 6: This is a reference, not rationale.

Author: @gbiggs thank you for your comment use "reference" here and above.

Operations
==========

1-A. Orphaning a package by maintainers
---------------------------------------

If you can no longer maintain a your package.

a. Set the package maintainer section in the ``package.xml`` file as follows.

   ``<maintainer email="ros-orphaned-packages@googlegroups.com">ROS Orphaned Package Maintainers</maintainer>``

> b. Send email to ros-orphaned-packages@googlegroups.com and/or post to the `ROS Discourse Release Category <http://discourse.ros.org/c/release>`_, indicating that package is now orphaned.

Reviewer 6: Have either a mailing list or (preferably) a Discourse category. We can subscribe to Discourse categories so there is no need for a separate mailing list. If a mailing list is needed to provide an email for the maintainer tag, then either find out if Discourse can provide that sort of feature or have just a mailing list rather than a Discourse category.

a. Contact the current maintainer, ask for transition to orphaned package and request access to source and release repository in order to make releases.

b. If the current maintainer is not responsible (a single day isn't enough to conclude that) and you still want to release the packages, there are two different paths forward depend on the "location" of the source and release repositories.

   1. If the repositories are hosted on GitHub within an `ros-` org unit, the admins of the org unit will usually grant the necessary access to these repositories.

   2. If nobody else beside the original author (which doesn't respond) has access, you should fork the repository.

c. Add ROS Orphaned Package Maintainers to the maintainer section of package.xml as follows.

   ``<maintainer email="ros-orphaned-packages@googlegroups.com">ROS Orphaned Package Maintainers</maintainer>``

2. Adopting packages
--------------------

Orphaned packages are seeking a new maintainer.
If you willing to step forward the package maintainer, more than just to release into new ROS distros, but respond to the issues, pull requests and so on.

> Contact to the ROS Orphaned Package Maintainers by sending email to ros-orphaned-packages@googlegroups.com and/or post to the `ROS Discourse Release Category <http://discourse.ros.org/c/release>`_.

Reviewer 6: Have either a mailing list or (preferably) a Discourse category. We can subscribe to Discourse categories so there is no need for a separate mailing list. If a mailing list is needed to provide an email for the maintainer tag, then either find out if Discourse can provide that sort of feature or have just a mailing list rather than a Discourse category.

Author: @gbiggs you're right. Send a message to mailing list is not a good idea over Discourse. Changed to `Post a message to the `ROS Discourse Release Category <http://discourse.ros.org/c/release>`_ indicating that package is now orphaned.`

3. Releasing orphaned packages
------------------------------

a. The package release process is as same as normal packages, so please follow bloom tutorial titled ``Releasing into a new ROS Distribution`` in ROS wiki documents [9]_ to release the orphaned package.

   1. Updated maintainer section in ``package.xml`` indicating that the package is orphaned and commit to the source tree, before you run the ``catkin_generate_changelog`` command.

   2. During ``bloom-release`` process, you will asked to set maintenance status as follows.
   Choose ``unmaintained`` . ::
      
::

  Would you like to add a maintenance status for this repository?' [Y/n] Y
   Please enter a maintenance status.
   Valid maintenance statuses:
   - developed: active development is in progress
   - maintained: no new development, but bug fixes and pull requests are addressed

>    - unmaintained: looking for new maintainer, bug fixes and pull requests will not be addressed

Reviewer 6: Should probably also mention that when it becomes incompatible with a dependency, it will no longer be released unless someone contributes a patch to fix it.
